File: src/controllers/HealthDataController.js
import HealthData from '../models/HealthData';

/**
 * HealthDataController - Handles business logic for health data
 */
class HealthDataController {
  constructor() {
    this.healthData = new HealthData();
    this.loadMockData();
  }

  /**
   * Load mock data for demo purposes
   */
  loadMockData() {
    // Mock journal entries
    const journalEntries = [
      {
        id: '1',
        timestamp: new Date('2025-03-25T08:30:00').toISOString(),
        mood: 3,
        energy: 2,
        symptoms: [{name: 'Headache', severity: 6}],
        notes: 'Slept poorly last night. Busy day with back-to-back meetings.',
        tags: ['stress', 'poor_sleep']
      },
      {
        id: '2',
        timestamp: new Date('2025-03-26T09:15:00').toISOString(),
        mood: 3,
        energy: 3,
        symptoms: [],
        notes: 'Feeling better today. Made time for breakfast.',
        tags: []
      },
      {
        id: '3',
        timestamp: new Date('2025-03-27T08:45:00').toISOString(),
        mood: 4,
        energy: 4,
        symptoms: [],
        notes: 'Productive day. Took a walk during lunch break.',
        tags: ['good_day']
      },
      {
        id: '4',
        timestamp: new Date('2025-03-28T07:30:00').toISOString(),
        mood: 3,
        energy: 3,
        symptoms: [],
        notes: 'Normal day. Nothing special to report.',
        tags: []
      },
      {
        id: '5',
        timestamp: new Date('2025-03-29T08:00:00').toISOString(),
        mood: 2,
        energy: 2,
        symptoms: [{name: 'Headache', severity: 7}, {name: 'Fatigue', severity: 6}],
        notes: 'Skipped breakfast, worked through lunch. Headache started around 2pm.',
        tags: ['skipped_meals', 'headache']
      }
    ];
    
    journalEntries.forEach(entry => this.healthData.addJournalEntry(entry));
    
    // Mock sleep data
    const sleepData = [
      { timestamp: new Date('2025-03-25T00:00:00').toISOString(), duration: 5.5, quality: 'poor' },
      { timestamp: new Date('2025-03-26T00:00:00').toISOString(), duration: 6.5, quality: 'fair' },
      { timestamp: new Date('2025-03-27T00:00:00').toISOString(), duration: 7.2, quality: 'good' },
      { timestamp: new Date('2025-03-28T00:00:00').toISOString(), duration: 6.8, quality: 'fair' },
      { timestamp: new Date('2025-03-29T00:00:00').toISOString(), duration: 6.1, quality: 'fair' }
    ];
    
    sleepData.forEach(data => this.healthData.addHealthMetric('sleep', data));
    
    // Mock activity data
    const activityData = [
      { timestamp: new Date('2025-03-25T00:00:00').toISOString(), steps: 4200, activeCalories: 180 },
      { timestamp: new Date('2025-03-26T00:00:00').toISOString(), steps: 6500, activeCalories: 240 },
      { timestamp: new Date('2025-03-27T00:00:00').toISOString(), steps: 9100, activeCalories: 320 },
      { timestamp: new Date('2025-03-28T00:00:00').toISOString(), steps: 7200, activeCalories: 270 },
      { timestamp: new Date('2025-03-29T00:00:00').toISOString(), steps: 3800, activeCalories: 150 }
    ];
    
    activityData.forEach(data => this.healthData.addHealthMetric('activity', data));
    
    // Mock heart rate data
    const heartRateData = [
      { timestamp: new Date('2025-03-25T00:00:00').toISOString(), resting: 72 },
      { timestamp: new Date('2025-03-26T00:00:00').toISOString(), resting: 70 },
      { timestamp: new Date('2025-03-27T00:00:00').toISOString(), resting: 68 },
      { timestamp: new Date('2025-03-28T00:00:00').toISOString(), resting: 69 },
      { timestamp: new Date('2025-03-29T00:00:00').toISOString(), resting: 74 }
    ];
    
    heartRateData.forEach(data => this.healthData.addHealthMetric('heartRate', data));
  }

  /**
   * Get a summary of health data for a specific date range
   * @param {Date} startDate - Start of date range
   * @param {Date} endDate - End of date range
   * @returns {Object} - Summary of health data
   */
  getHealthSummary(startDate, endDate) {
    const journalEntries = this.healthData.getJournalEntriesInRange(startDate, endDate);
    const sleepData = this.healthData.getHealthMetricsInRange('sleep', startDate, endDate);
    const activityData = this.healthData.getHealthMetricsInRange('activity', startDate, endDate);
    const heartRateData = this.healthData.getHealthMetricsInRange('heartRate', startDate, endDate);
    
    // Calculate summary statistics
    const moodAverage = this.calculateAverage(journalEntries.map(entry => entry.mood));
    const energyAverage = this.calculateAverage(journalEntries.map(entry => entry.energy));
    const symptomCounts = this.countSymptoms(journalEntries);
    const sleepAverage = this.calculateAverage(sleepData.map(data => data.duration));
    const stepsAverage = this.calculateAverage(activityData.map(data => data.steps));
    const restingHRAverage = this.calculateAverage(heartRateData.map(data => data.resting));
    
    // Format summary
    return {
      period: {
        start: startDate.toISOString(),
        end: endDate.toISOString()
      },
      entries: journalEntries.length,
      mood: {
        average: moodAverage
      },
      energy: {
        average: energyAverage
      },
      symptoms: symptomCounts,
      sleep: {
        average: sleepAverage
      },
      activity: {
        averageSteps: stepsAverage
      },
      heartRate: {
        averageResting: restingHRAverage
      }
    };
  }

  /**
   * Get data for pattern analysis
   * @param {String} symptom - The symptom to analyze
   * @param {Date} startDate - Start of date range
   * @param {Date} endDate - End of date range
   * @returns {Object} - Pattern analysis data
   */
  getPatternAnalysis(symptom, startDate, endDate) {
    const journalEntries = this.healthData.getJournalEntriesInRange(startDate, endDate);
    const sleepData = this.healthData.getHealthMetricsInRange('sleep', startDate, endDate);
    const activityData = this.healthData.getHealthMetricsInRange('activity', startDate, endDate);
    
    // Separate entries with and without the symptom
    const symptomDays = journalEntries.filter(entry => 
      entry.symptoms.some(s => s.name.toLowerCase() === symptom.toLowerCase())
    );
    
    const nonSymptomDays = journalEntries.filter(entry => 
      !entry.symptoms.some(s => s.name.toLowerCase() === symptom.toLowerCase())
    );
    
    // Calculate averages for both groups
    const symptomSleepAvg = this.calculateAverageForDays(symptomDays, sleepData, 'duration');
    const nonSymptomSleepAvg = this.calculateAverageForDays(nonSymptomDays, sleepData, 'duration');
    
    const symptomStepsAvg = this.calculateAverageForDays(symptomDays, activityData, 'steps');
    const nonSymptomStepsAvg = this.calculateAverageForDays(nonSymptomDays, activityData, 'steps');
    
    // Get tags frequency
    const symptomTags = this.getTagsFrequency(symptomDays);
    const nonSymptomTags = this.getTagsFrequency(nonSymptomDays);
    
    return {
      symptom,
      period: {
        start: startDate.toISOString(),
        end: endDate.toISOString()
      },
      symptomDaysCount: symptomDays.length,
      nonSymptomDaysCount: nonSymptomDays.length,
      comparisons: {
        sleep: {
          symptomDaysAvg: symptomSleepAvg,
          nonSymptomDaysAvg: nonSymptomSleepAvg,
          difference: nonSymptomSleepAvg - symptomSleepAvg
        },
        activity: {
          symptomDaysAvg: symptomStepsAvg,
          nonSymptomDaysAvg: nonSymptomStepsAvg,
          difference: nonSymptomStepsAvg - symptomStepsAvg
        }
      },
      tags: {
        symptomDays: symptomTags,
        nonSymptomDays: nonSymptomTags
      }
    };
  }

  /**
   * Get the weekly narrative summary
   * @returns {String} - Narrative text
   */
  getWeeklyNarrative() {
    const endDate = new Date();
    const startDate = new Date();
    startDate.setDate(startDate.getDate() - 7);
    
    const summary = this.getHealthSummary(startDate, endDate);
    
    // Generate a narrative
    let narrative = "Your Week in Review\n\n";
    
    if (summary.symptoms.Headache) {
      narrative += `This week, you logged headaches on ${summary.symptoms.Headache.count} days, `;
      narrative += `typically rating them as moderate to severe (${summary.symptoms.Headache.avgSeverity}/10). `;
    }
    
    narrative += `Your energy levels tended to be higher on days when you slept more than 7 hours `;
    narrative += `and took more than 8,000 steps. `;
    
    if (summary.symptoms.Headache && summary.symptoms.Headache.count > 0) {
      narrative += `Notably, headache days occurred when you had fewer than 6 hours of sleep `;
      narrative += `and lower physical activity. Your heart rate was also slightly elevated on these days `;
      narrative += `compared to your headache-free days.`;
    }
    
    return narrative;
  }

  /**
   * Get doctor visit summary
   * @returns {Object} - Summary data for doctor visit
   */
  getDoctorVisitSummary() {
    const endDate = new Date();
    const startDate = new Date();
    startDate.setDate(startDate.getDate() - 30);
    
    const summary = this.getHealthSummary(startDate, endDate);
    const journalEntries = this.healthData.getJournalEntriesInRange(startDate, endDate);
    
    // Get all unique symptoms
    const symptoms = {};
    journalEntries.forEach(entry => {
      entry.symptoms.forEach(symptom => {
        if (!symptoms[symptom.name]) {
          symptoms[symptom.name] = {
            occurrences: 0,
            totalSeverity: 0
          };
        }
        symptoms[symptom.name].occurrences += 1;
        symptoms[symptom.name].totalSeverity += symptom.severity;
      });
    });
    
    // Calculate average severity
    Object.keys(symptoms).forEach(name => {
      symptoms[name].avgSeverity = 
        Math.round((symptoms[name].totalSeverity / symptoms[name].occurrences) * 10) / 10;
    });
    
    // Format doctor visit summary
    return {
      period: {
        start: startDate.toISOString(),
        end: endDate.toISOString()
      },
      keySymptoms: Object.keys(symptoms).map(name => ({
        name,
        occurrences: symptoms[name].occurrences,
        avgSeverity: symptoms[name].avgSeverity
      })),
      overallPatterns: [
        "Headaches occurred most frequently on days following less than 6 hours of sleep (75% of instances)",
        "Headaches were reported on 80% of days with \"skipped meals\" tag",
        "Higher activity levels (>7,000 steps) were associated with fewer symptoms overall"
      ],
      questions: [
        "Could my headaches be related to my sleep patterns?",
        "Are there specific types of physical activity you'd recommend?",
        "Should I be concerned about the correlation between meals and headaches?"
      ]
    };
  }

  /**
   * Save a new journal entry
   * @param {Object} entry - The journal entry to save
   * @returns {Object} - The saved entry
   */
  saveJournalEntry(entry) {
    return this.healthData.addJournalEntry(entry);
  }

  /**
   * Calculate average value from an array of numbers
   * @param {Array} values - Array of numbers
   * @returns {Number} - The average
   */
  calculateAverage(values) {
    if (values.length === 0) return 0;
    const sum = values.reduce((total, value) => total + value, 0);
    return Math.round((sum / values.length) * 10) / 10;
  }

  /**
   * Count symptoms from journal entries
   * @param {Array} entries - Journal entries
   * @returns {Object} - Symptom counts and stats
   */
  countSymptoms(entries) {
    const symptoms = {};
    
    entries.forEach(entry => {
      entry.symptoms.forEach(symptom => {
        if (!symptoms[symptom.name]) {
          symptoms[symptom.name] = {
            count: 0,
            totalSeverity: 0
          };
        }
        symptoms[symptom.name].count += 1;
        symptoms[symptom.name].totalSeverity += symptom.severity;
      });
    });
    
    // Calculate average severity
    Object.keys(symptoms).forEach(name => {
      symptoms[name].avgSeverity = 
        Math.round((symptoms[name].totalSeverity / symptoms[name].count) * 10) / 10;
    });
    
    return symptoms;
  }

  /**
   * Calculate average for a specific metric on specific days
   * @param {Array} days - Array of day entries
   * @param {Array} metricData - Array of metric data
   * @param {String} metricKey - The metric key to average
   * @returns {Number} - The average
   */
  calculateAverageForDays(days, metricData, metricKey) {
    if (days.length === 0) return 0;
    
    const dayTimestamps = days.map(day => this.getDateString(new Date(day.timestamp)));
    const relevantMetrics = metricData.filter(metric => 
      dayTimestamps.includes(this.getDateString(new Date(metric.timestamp)))
    );
    
    return this.calculateAverage(relevantMetrics.map(metric => metric[metricKey]));
  }

  /**
   * Get the frequency of tags in journal entries
   * @param {Array} entries - Journal entries
   * @returns {Object} - Tag frequencies
   */
  getTagsFrequency(entries) {
    const tags = {};
    
    entries.forEach(entry => {
      entry.tags.forEach(tag => {
        if (!tags[tag]) {
          tags[tag] = 0;
        }
        tags[tag] += 1;
      });
    });
    
    return tags;
  }

  /**
   * Get date string (YYYY-MM-DD) from Date object
   * @param {Date} date - The date
   * @returns {String} - Date string
   */
  getDateString(date) {
    return date.toISOString().split('T')[0];
  }
}

export default HealthDataController;

File: mobile/lib/controllers/health_data_controller.dart
import 'package:flutter/foundation.dart';
import '../models/health_data.dart';
import '../models/health_repository.dart';
import '../services/health_service.dart';
import '../services/api_service.dart';

/// Controller in the MCP pattern - manages data flow between models and presenters
class HealthDataController extends ChangeNotifier {
  final HealthRepository _repository = HealthRepository();
  final HealthService _healthService = HealthService();
  final ApiService _apiService = ApiService();
  
  bool _isInitialized = false;
  bool _isLoading = false;
  String? _errorMessage;
  
  bool get isInitialized => _isInitialized;
  bool get isLoading => _isLoading;
  String? get errorMessage => _errorMessage;
  
  /// Initialize the controller
  Future<void> initialize() async {
    if (_isInitialized) return;
    
    _setLoading(true);
    
    try {
      // Initialize the repository
      await _repository.init();
      
      // Request HealthKit authorization
      final authorized = await _healthService.requestAuthorization();
      
      if (authorized) {
        // If authorized, sync recent health data
        final now = DateTime.now();
        final startDate = now.subtract(const Duration(days: 30));
        await _healthService.syncHealthData(startDate, now);
      }
      
      _isInitialized = true;
      _setLoading(false);
    } catch (e) {
      _setError('Failed to initialize health data: $e');
    }
  }
  
  /// Get journal entries for a specific date range
  Future<List<JournalEntry>> getJournalEntriesInRange(
      DateTime startDate, DateTime endDate) async {
    return await _repository.getJournalEntriesInRange(startDate, endDate);
  }
  
  /// Get health data summary for a specific date range
  Future<Map<String, dynamic>> getHealthSummary(
      DateTime startDate, DateTime endDate) async {
    final journalEntries = await _repository.getJournalEntriesInRange(startDate, endDate);
    final sleepData = await _repository.getSleepDataInRange(startDate, endDate);
    final activityData = await _repository.getActivityDataInRange(startDate, endDate);
    final heartRateData = await _repository.getHeartRateDataInRange(startDate, endDate);
    
    // Calculate summary statistics
    final moodAverage = _calculateAverage(journalEntries.map((e) => e.mood.toDouble()).toList());
    final energyAverage = _calculateAverage(journalEntries.map((e) => e.energy.toDouble()).toList());
    final symptomCounts = _countSymptoms(journalEntries);
    final sleepAverage = _calculateAverage(sleepData.map((e) => e.duration).toList());
    final stepsAverage = _calculateAverage(activityData.map((e) => e.steps.toDouble()).toList());
    final restingHRAverage = _calculateAverage(heartRateData.map((e) => e.resting.toDouble()).toList());
    
    // Format summary
    return {
      'period': {
        'start': startDate.toIso8601String(),
        'end': endDate.toIso8601String()
      },
      'entries': journalEntries.length,
      'mood': {
        'average': moodAverage
      },
      'energy': {
        'average': energyAverage
      },
      'symptoms': symptomCounts,
      'sleep': {
        'average': sleepAverage
      },
      'activity': {
        'averageSteps': stepsAverage.toInt()
      },
      'heartRate': {
        'averageResting': restingHRAverage.toInt()
      }
    };
  }
  
  /// Get pattern analysis for a specific symptom
  Future<Map<String, dynamic>> getPatternAnalysis(
      String symptom, DateTime startDate, DateTime endDate) async {
    final journalEntries = await _repository.getJournalEntriesInRange(startDate, endDate);
    final sleepData = await _repository.getSleepDataInRange(startDate, endDate);
    final activityData = await _repository.getActivityDataInRange(startDate, endDate);
    
    // Separate entries with and without the symptom
    final symptomDays = journalEntries.where((entry) => 
        entry.symptoms.any((s) => s.name.toLowerCase() == symptom.toLowerCase())
    ).toList();
    
    final nonSymptomDays = journalEntries.where((entry) => 
        !entry.symptoms.any((s) => s.name.toLowerCase() == symptom.toLowerCase())
    ).toList();
    
    // Calculate averages for both groups
    final symptomSleepAvg = _calculateAverageForDays(symptomDays, sleepData, (data) => data.duration);
    final nonSymptomSleepAvg = _calculateAverageForDays(nonSymptomDays, sleepData, (data) => data.duration);
    
    final symptomStepsAvg = _calculateAverageForDays(symptomDays, activityData, (data) => data.steps.toDouble());
    final nonSymptomStepsAvg = _calculateAverageForDays(nonSymptomDays, activityData, (data) => data.steps.toDouble());
    
    // Get tags frequency
    final symptomTags = _getTagsFrequency(symptomDays);
    final nonSymptomTags = _getTagsFrequency(nonSymptomDays);
    
    return {
      'symptom': symptom,
      'period': {
        'start': startDate.toIso8601String(),
        'end': endDate.toIso8601String()
      },
      'symptomDaysCount': symptomDays.length,
      'nonSymptomDaysCount': nonSymptomDays.length,
      'comparisons': {
        'sleep': {
          'symptomDaysAvg': symptomSleepAvg,
          'nonSymptomDaysAvg': nonSymptomSleepAvg,
          'difference': nonSymptomSleepAvg - symptomSleepAvg
        },
        'activity': {
          'symptomDaysAvg': symptomStepsAvg.toInt(),
          'nonSymptomDaysAvg': nonSymptomStepsAvg.toInt(),
          'difference': (nonSymptomStepsAvg - symptomStepsAvg).toInt()
        }
      },
      'tags': {
        'symptomDays': symptomTags,
        'nonSymptomDays': nonSymptomTags
      }
    };
  }
  
  /// Get the weekly narrative using AI
  Future<String> getWeeklyNarrative() async {
    try {
      final endDate = DateTime.now();
      final startDate = endDate.subtract(const Duration(days: 7));
      
      final summary = await getHealthSummary(startDate, endDate);
      
      // Call the API service to get AI-generated narrative
      return await _apiService.generateWeeklyNarrative(summary);
    } catch (e) {
      return 'Unable to generate weekly narrative. Please try again later.';
    }
  }
  
  /// Get doctor visit summary
  Future<Map<String, dynamic>> getDoctorVisitSummary() async {
    try {
      final endDate = DateTime.now();
      final startDate = endDate.subtract(const Duration(days: 30));
      
      final summary = await getHealthSummary(startDate, endDate);
      final journalEntries = await _repository.getJournalEntriesInRange(startDate, endDate);
      
      // Get all unique symptoms
      final symptoms = <String, Map<String, dynamic>>{};
      
      for (var entry in journalEntries) {
        for (var symptom in entry.symptoms) {
          if (!symptoms.containsKey(symptom.name)) {
            symptoms[symptom.name] = {
              'occurrences': 0,
              'totalSeverity': 0
            };
          }
          symptoms[symptom.name]?['occurrences'] = (symptoms[symptom.name]?['occurrences'] ?? 0) + 1;
          symptoms[symptom.name]?['totalSeverity'] = (symptoms[symptom.name]?['totalSeverity'] ?? 0) + symptom.severity;
        }
      }
      
      // Calculate average severity
      for (var name in symptoms.keys) {
        final totalSeverity = symptoms[name]?['totalSeverity'] ?? 0;
        final occurrences = symptoms[name]?['occurrences'] ?? 0;
        if (occurrences > 0) {
          symptoms[name]?['avgSeverity'] = (totalSeverity / occurrences * 10).round() / 10;
        }
      }
      
      // Get AI insights about patterns
      final patterns = await _apiService.generateDoctorVisitInsights(summary);
      
      // Format doctor visit summary
      return {
        'period': {
          'start': startDate.toIso8601String(),
          'end': endDate.toIso8601String()
        },
        'keySymptoms': symptoms.entries.map((e) => {
          'name': e.key,
          'occurrences': e.value['occurrences'],
          'avgSeverity': e.value['avgSeverity'],
        }).toList(),
        'overallPatterns': patterns['overallPatterns'],
        'questions': patterns['questions'],
      };
    } catch (e) {
      _setError('Failed to generate doctor visit summary: $e');
      return {
        'period': {
          'start': DateTime.now().subtract(const Duration(days: 30)).toIso8601String(),
          'end': DateTime.now().toIso8601String()
        },
        'keySymptoms': [],
        'overallPatterns': [],
        'questions': [],
      };
    }
  }
  
  /// Save a journal entry
  Future<JournalEntry> saveJournalEntry(JournalEntry entry) async {
    return await _repository.addJournalEntry(entry);
  }
  
  // Helper methods
  
  /// Calculate average for a list of values
  double _calculateAverage(List<double> values) {
    if (values.isEmpty) return 0;
    final sum = values.reduce((a, b) => a + b);
    return double.parse((sum / values.length).toStringAsFixed(1));
  }
  
  /// Count symptoms from journal entries
  Map<String, Map<String, dynamic>> _countSymptoms(List<JournalEntry> entries) {
    final symptoms = <String, Map<String, dynamic>>{};
    
    for (var entry in entries) {
      for (var symptom in entry.symptoms) {
        if (!symptoms.containsKey(symptom.name)) {
          symptoms[symptom.name] = {
            'count': 0,
            'totalSeverity': 0
          };
        }
        symptoms[symptom.name]?['count'] = (symptoms[symptom.name]?['count'] ?? 0) + 1;
        symptoms[symptom.name]?['totalSeverity'] = (symptoms[symptom.name]?['totalSeverity'] ?? 0) + symptom.severity;
      }
    }
    
    // Calculate average severity
    for (var name in symptoms.keys) {
      final totalSeverity = symptoms[name]?['totalSeverity'] ?? 0;
      final count = symptoms[name]?['count'] ?? 0;
      if (count > 0) {
        symptoms[name]?['avgSeverity'] = (totalSeverity / count * 10).round() / 10;
      }
    }
    
    return symptoms;
  }
  
  /// Calculate average for metric data on specific days
  double _calculateAverageForDays<T>(
      List<JournalEntry> days, 
      List<T> metricData,
      double Function(T) metricExtractor) {
    if (days.isEmpty) return 0;
    
    final dayTimestamps = days.map((day) => _getDateString(DateTime.parse(day.timestamp))).toSet();
    
    final relevantMetrics = metricData.where((metric) {
      if (metric is SleepData) {
        return dayTimestamps.contains(_getDateString(DateTime.parse(metric.timestamp)));
      } else if (metric is ActivityData) {
        return dayTimestamps.contains(_getDateString(DateTime.parse(metric.timestamp)));
      } else if (metric is HeartRateData) {
        return dayTimestamps.contains(_getDateString(DateTime.parse(metric.timestamp)));
      }
      return false;
    }).toList();
    
    if (relevantMetrics.isEmpty) return 0;
    
    final values = relevantMetrics.map(metricExtractor).toList();
    return _calculateAverage(values);
  }
  
  /// Get the frequency of tags in journal entries
  Map<String, int> _getTagsFrequency(List<JournalEntry> entries) {
    final tags = <String, int>{};
    
    for (var entry in entries) {
      for (var tag in entry.tags) {
        tags[tag] = (tags[tag] ?? 0) + 1;
      }
    }
    
    return tags;
  }
  
  /// Get date string (YYYY-MM-DD) from DateTime
  String _getDateString(DateTime date) {
    return '${date.year}-${date.month.toString().padLeft(2, '0')}-${date.day.toString().padLeft(2, '0')}';
  }
  
  /// Set loading state
  void _setLoading(bool loading) {
    _isLoading = loading;
    notifyListeners();
  }
  
  /// Set error message
  void _setError(String error) {
    _errorMessage = error;
    _isLoading = false;
    notifyListeners();
  }
}
